File: android/app/src/main/java/com/zebwqfox/hbustpower/model/PowerThemeStyle.kt
package com.zebwqfox.hbustpower.model

/**
 * Colour schemes (iOS 1.9.0 `PowerThemeStyle`). The two extra ones are modelled on friends' fursuits:
 * 紫鸟紫 (black fur, violet markings, amber eyes) and 枫烻黄 (grey fur, yellow markings, red collar with a bell).
 * Colours are stored as 0xAARRGGBB so this stays free of Compose types and can be unit tested.
 */
enum class PowerThemeStyle(
    val id: String,
    val displayName: String,
    val tagline: String,
    /** Where the colours come from; null for the app's own scheme. */
    val credit: String?,
    val greeting: String,
    /** An extra line the pull-to-charge indicator may show in this theme. */
    val chargingLine: String,
    val light: Palette,
    val dark: Palette,
) {
    CLASSIC(
        id = "classic",
        displayName = "湖科蓝",
        tagline = "云白配清蓝，App 原本的样子",
        credit = null,
        greeting = "⚡ 电力满满",
        chargingLine = "电流正在赶来…",
        light = Palette(0xFFF6F9FE, 0xFFFFFFFF, 0xFF1F57C2, 0xFFDBECFF, 0xFFAB6621, 0xFFFFA80D),
        dark = Palette(0xFF0A0F1A, 0xFF141C2B, 0xFF85B8FF, 0xFF16294A, 0xFFF0BA66, 0xFFFFCC4D),
    ),
    PURPLE_BIRD(
        id = "purpleBird",
        displayName = "紫鸟紫",
        tagline = "夜色里的紫与琥珀色的眼睛",
        credit = "配色来自朋友「紫鸟紫」",
        greeting = "🌙 夜里也在偷偷发电",
        chargingLine = "夜色里悄悄充电…",
        light = Palette(0xFFF9F7FE, 0xFFFFFFFF, 0xFF5C2EC7, 0xFFE6DEFC, 0xFFC7730D, 0xFF8C4DF2),
        dark = Palette(0xFF0E0B16, 0xFF1D182A, 0xFFB89BFF, 0xFF291F4D, 0xFFFFC259, 0xFFCCA8FF),
    ),
    MAPLE_YELLOW(
        id = "mapleYellow",
        displayName = "枫烻黄",
        tagline = "灰黄配色，脖子上挂着金铃铛",
        credit = "配色来自朋友「枫烻黄」",
        greeting = "🔔 叮——铃铛响了一下",
        chargingLine = "铃铛一响，电就来了…",
        light = Palette(0xFFFEFBF5, 0xFFFFFFFF, 0xFF996B05, 0xFFFFF0C7, 0xFFB82921, 0xFFF2B30D),
        dark = Palette(0xFF131108, 0xFF252117, 0xFFFFD147, 0xFF382E12, 0xFFFF806B, 0xFFFFD94D),
    ),
    ;

    /** One appearance of a theme; only the colours that actually change between themes live here. */
    data class Palette(
        val background: Long,
        val surface: Long,
        val accent: Long,
        /** The balance card's tint. */
        val hero: Long,
        /** The warm secondary: lighting figures, low-balance warnings and stickers. */
        val secondary: Long,
        /** Sparks and the charging ring; always bright enough to read as "electric". */
        val spark: Long,
    )

    companion object {
        val default = CLASSIC
        fun of(id: String?): PowerThemeStyle = entries.firstOrNull { it.id == id } ?: default
    }
}

/**
 * Lines from Tagore's *Stray Birds* (1916) in 郑振铎's 1922 Chinese translation; both are out of copyright.
 * Shown once per launch in the 紫鸟紫 theme, always with the attribution.
 */
object StrayBirds {
    const val ATTRIBUTION = "摘自《飞鸟集》· 泰戈尔"

    val lines = listOf(
        "夏天的飞鸟，飞到我的窗前唱歌，又飞去了。",
        "世界上的一队小小的漂泊者呀，请留下你们的足印在我的文字里。",
        "如果你因失去了太阳而流泪，那么你也将失去群星了。",
        "我今晨坐在窗前，世界如一个路人似的，停留了一会，向我点点头又走过去了。",
        "休息与工作的关系，正如眼睑与眼睛的关系。",
        "瀑布歌唱道：我得到自由时便有了歌声了。",
        "我的心把她的波浪在世界的海岸上冲激着，以热泪在上边写着她的题记：我爱你。",
        "群星不怕显得像萤火那样。",
        "使生如夏花之绚烂，死如秋叶之静美。",
        "这寡独的黄昏，幕着雾与雨，我在我的心的孤寂里，感觉到它的叹息。",
        "我们把世界看错了，反说它欺骗我们。",
        "夜秘密地把花开放了，却让白日去领受谢词。",
        "小草呀，你的足步虽小，但是你拥有你足下的土地。",
        "错误经不起失败，但是真理却不怕失败。",
        "只管走过去，不必逗留着采了花朵来保存，因为一路上花朵自会继续开放的。",
    )

    /** One line per launch, so the home screen greets you with something new each time you open the app. */
    val ofThisLaunch: String by lazy { lines.random() }

    /** Used when the reader taps for another one. */
    fun after(current: String): String {
        val index = lines.indexOf(current)
        return if (index < 0) ofThisLaunch else lines[(index + 1) % lines.size]
    }
}

File: ios/Shared/PowerThemeStyle.swift
import UIKit

/// Colour schemes. The two extra ones are modelled on friends' fursuits: 紫鸟紫 (black fur, violet markings,
/// amber eyes) and 枫烻黄 (grey fur, yellow markings, red collar with a gold bell).
enum PowerThemeStyle: String, CaseIterable, Sendable {
    case classic, purpleBird, mapleYellow

    var name: String {
        switch self {
        case .classic: "湖科蓝"
        case .purpleBird: "紫鸟紫"
        case .mapleYellow: "枫烻黄"
        }
    }

    var tagline: String {
        switch self {
        case .classic: "云白配清蓝，App 原本的样子"
        case .purpleBird: "夜色里的紫与琥珀色的眼睛"
        case .mapleYellow: "灰黄配色，脖子上挂着金铃铛"
        }
    }

    /// Credit shown under the friends' themes.
    var credit: String? {
        switch self {
        case .classic: nil
        case .purpleBird, .mapleYellow: "配色来自朋友「\(name)」"
        }
    }

    var symbol: String {
        switch self {
        case .classic: "bolt.fill"
        case .purpleBird: "moon.stars.fill"
        case .mapleYellow: "bell.fill"
        }
    }

    /// Shown when the theme's signature animation plays.
    var greeting: String {
        switch self {
        case .classic: "⚡ 电力满满"
        case .purpleBird: "🌙 夜里也在偷偷发电"
        case .mapleYellow: "🔔 叮——铃铛响了一下"
        }
    }

    /// An extra line the pull-to-charge indicator may show in this theme.
    var chargingLine: String {
        switch self {
        case .classic: "电流正在赶来…"
        case .purpleBird: "夜色里悄悄充电…"
        case .mapleYellow: "铃铛一响，电就来了…"
        }
    }

    // MARK: Colours

    fileprivate struct Pair {
        let light: UIColor
        let dark: UIColor
        var dynamic: UIColor { UIColor { $0.userInterfaceStyle == .dark ? dark : light } }
    }

    private var backgroundPair: Pair {
        switch self {
        case .classic: Pair(light: .rgb(0.965, 0.977, 0.995), dark: .rgb(0.04, 0.06, 0.10))
        case .purpleBird: Pair(light: .rgb(0.975, 0.970, 0.995), dark: .rgb(0.055, 0.042, 0.085))
        case .mapleYellow: Pair(light: .rgb(0.995, 0.985, 0.960), dark: .rgb(0.075, 0.068, 0.045))
        }
    }

    private var surfacePair: Pair {
        switch self {
        case .classic: Pair(light: .white, dark: .rgb(0.08, 0.11, 0.17))
        case .purpleBird: Pair(light: .white, dark: .rgb(0.115, 0.095, 0.165))
        case .mapleYellow: Pair(light: .white, dark: .rgb(0.145, 0.130, 0.090))
        }
    }

    private var accentPair: Pair {
        switch self {
        case .classic: Pair(light: .rgb(0.12, 0.34, 0.76), dark: .rgb(0.52, 0.72, 1.0))
        case .purpleBird: Pair(light: .rgb(0.36, 0.18, 0.78), dark: .rgb(0.72, 0.61, 1.0))
        case .mapleYellow: Pair(light: .rgb(0.60, 0.42, 0.02), dark: .rgb(1.0, 0.82, 0.28))
        }
    }

    private var heroPair: Pair {
        switch self {
        case .classic: Pair(light: .rgb(0.86, 0.925, 1.0), dark: .rgb(0.085, 0.16, 0.29))
        case .purpleBird: Pair(light: .rgb(0.90, 0.87, 0.99), dark: .rgb(0.16, 0.12, 0.30))
        case .mapleYellow: Pair(light: .rgb(1.0, 0.94, 0.78), dark: .rgb(0.22, 0.18, 0.07))
        }
    }

    /// The warm secondary: lighting figures, low-balance warnings and stickers.
    private var secondaryPair: Pair {
        switch self {
        case .classic: Pair(light: .rgb(0.67, 0.40, 0.13), dark: .rgb(0.94, 0.73, 0.40))
        case .purpleBird: Pair(light: .rgb(0.78, 0.45, 0.05), dark: .rgb(1.0, 0.76, 0.35))
        case .mapleYellow: Pair(light: .rgb(0.72, 0.16, 0.13), dark: .rgb(1.0, 0.50, 0.42))
        }
    }

    /// Sparks and the charging ring; always bright enough to read as "electric".
    private var sparkPair: Pair {
        switch self {
        case .classic: Pair(light: .rgb(1.0, 0.66, 0.05), dark: .rgb(1.0, 0.80, 0.30))
        case .purpleBird: Pair(light: .rgb(0.55, 0.30, 0.95), dark: .rgb(0.80, 0.66, 1.0))
        case .mapleYellow: Pair(light: .rgb(0.95, 0.70, 0.05), dark: .rgb(1.0, 0.85, 0.30))
        }
    }

    var background: UIColor { backgroundPair.dynamic }
    var surface: UIColor { surfacePair.dynamic }
    var accent: UIColor { accentPair.dynamic }
    var hero: UIColor { heroPair.dynamic }
    var secondary: UIColor { secondaryPair.dynamic }
    var spark: UIColor { sparkPair.dynamic }
}

private extension UIColor {
    static func rgb(_ red: CGFloat, _ green: CGFloat, _ blue: CGFloat) -> UIColor {
        UIColor(red: red, green: green, blue: blue, alpha: 1)
    }
}

/// Remembers the chosen theme. Changing it rebuilds the interface, since colours already drawn cannot repaint.
/// Only ever touched from the main thread, like the rest of the UI layer.
final class PowerThemeStore {
    nonisolated(unsafe) static let shared = PowerThemeStore()
    static let didChange = Notification.Name("powerThemeDidChange")
    private static let key = "appTheme"

    private let defaults: UserDefaults
    private(set) var style: PowerThemeStyle

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        style = (defaults.string(forKey: Self.key).flatMap(PowerThemeStyle.init(rawValue:))) ?? .classic
    }

    func select(_ style: PowerThemeStyle) {
        guard style != self.style else { return }
        self.style = style
        defaults.set(style.rawValue, forKey: Self.key)
        // Keep the widgets in the same colours as the app.
        PowerWidgetStore.saveTheme(style)
        NotificationCenter.default.post(name: Self.didChange, object: self)
    }
}
